blog/2019-03-28-debezium-at-pipedrive.adoc

```diff
@@ -0,0 +1,225 @@
+
+Our main alerts are based on https://debezium.io/docs/connectors/mysql/#binlog-metrics[MySQL_binlog_connected metric], but unfortunately it hasn’t proven absolutely reliable, so there is additional alerts based on logs, counting “org.apache.kafka.connect.errors.ConnectException” exceptions.
+
+Our grafana dashboard looks like this
+++++
+<div class="imageblock centered-image">
+    <img src="/images/pipedrive-grafana.png" style="max-width:100%; margin-bottom:20px; margin-top:20px;" class="responsive-image" alt="Debezium extended docker image">
+</div>
+++++
+
+== Problems and workarounds
+
+=== Databases with super large schemas.
+Debezium recommends giving connector 256MB memory for database with schema up to 10000 columns, but some of our servers have ~1.2 million columns! Debezium does work, but requires huge amount of memory to accommodate the in-memory schema. Also some operations that normally take  seconds start to crawl. For example database initial (schema only!) snapshot can run 15+ minutes and fill the history topic with half million “drop table”, “create table” messages. This in turn slows down connector restarts as all those database alters need to be played back to reconstruct correct in memory state.
+
+So what we’ve done so far (to make it perform acceptable level) is apply strict table whitelist for tracking 25% of the most critical tables. For large history topic configured connector to only contain whitelisted table DDLs. Also increasing java heap size to ~2GB has so far proven reliable.
+
+[source,json]
+----
+database.history.store.only.monitored.tables.ddl: "true",
+database.history.skip.unparseable.ddl": "true",
+----
+
+=== Snapshots without locking
+You can configure connector to use locking.mode=’none’, but it’s dangerous as ALTERS running during the snapshot will not be reflected in the Debezium in-memory database schema. You will learn about it only when the first data change event arrives for that table, which could be weeks after snapshot. So far the fix that worked has been to run schema recovery asap.
+
+== Lessons learned
+*Kafka connect in distributed mode is unstable*. Kafka-connect cluster doesn’t play well with large connectors count and slow task startup time - one connector restart can cause rebalance event, which means one repeatedly failing MySQL connector can halt whole cluster.
+
+*Manual Debezium offsets adjusting has saved us many times* - storing offsets in external kafka topic allowed us to easily adjust connector’s last offset by producing valid json message and restarting container. With active-passive setup unfortunate server failure can introduce new gaps in gtid ranges or slave or debezium itself can become ahead of new master on particular gtid channel. Fixing this without reseting connector needs manual intervention to last offset.
```

Comment:
```suggestion
*Manual Debezium offsets adjusting has saved us many times* - storing offsets in external Kafka topic allowed us to easily adjust connector’s last offset by producing a valid JSON message and restarting the container. With active-passive setup, an unfortunate server failure can introduce new gaps in GTID ranges, or the slave or Debezium itself can get ahead of the new master on a particular GTID channel. Fixing this without resetting connector needs manual intervention to last offset.
```

On this one, is there anything we can do about it? If so, could you file an in issue in JIRA?

Reply:
Yes, I'll add issue.

Problem is in mysql side, but debezium could help getting around it.

Some examples: 
1) master-slave setup, each own gtid channel + master and slave also has each others (failover has happened once).

Now even when slave is read-only mode some replication events can add changes to slave gtid channel. In memory tables in particular. 
*the first time that the master uses a given MEMORY table after startup, it logs an event that notifies slaves that the table must to be emptied by writing a DELETE statement for that table to the binary log. MySQL will still always write the DELETE query after a server restart, even if it’s for a table not to be replicated.*(https://dev.mysql.com/doc/refman/5.7/en/replication-features-memory.html).

What happens is slave own gtid channel will be ahead of what is in master, and those changes will be removed from binlog after some days.

Now when failover happens, debezium finds that the offsets for this gtid channel are behind of what's in new master (old slave), but master also has already purged them! So replication fails, until manually offset pushed (usually by less than 10 steps).

2) MySQL master server crashing hard (segfault) - will introduce new availability holes in gtid channel.

For example this: 
c0ab881b-31ec-11e8-a311-fa163ea867bf:1-44725:44727-45214:45216-52812:52814-54900:54902-505192,

For debezium this is already blocker, even though it should only care about events after it's last offset position (505192) in this case. Again fix right now is to manually write debezium-offsets topic message with new gtid holes, then connector can resume.

3) Debezium getting "ahead" of new master after failover. As our debezium is connector to mysql master, it's possible that async slave instance has huge lag and debezium commits offsets ahead of what mysql slave has processed. When on that moment failover happens, there's small data loss in mysql servers and debezium has offsets in master gtid channel ahead of new master. This is again blocker and manually fixed by pushing dbz offsets back. This is very rare, but with large number servers things do happen.

Reply:
created issue https://issues.jboss.org/browse/DBZ-1200